execUtil.go

```diff
@@ -1,62 +1,150 @@
+	statusResponse.ExecError = command.GetExecError()
+	statusResponse.Finished = new(bool)
+	if command.GetEndTime() != nil {
+		*statusResponse.Finished = true
+		statusResponse.EndTime = new(time.Time)
+		statusResponse.EndTime = command.GetEndTime()
+		statusResponse.ExitCode = new(int)
+		*statusResponse.ExitCode = command.GetExitCode()
+	}
+
+	statusResponse.Output = command.StdoutBuff.Lines()
+
+	return &statusResponse, nil
+}
+
+// RunCommand starts a process from command argument.
+func RunCommand(command string) (*models.StartResponse, error) {
+
+	var startResponse models.StartResponse
+	splitCommand := strings.Split(command, " ")
+
+	cmd := exec.Command(splitCommand[0], splitCommand[1:]...)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	outBuf := NewOutput()
+	cmd.Stdout = io.MultiWriter(os.Stdout, outBuf)
+
+	err := cmd.Start()
 
-  go func() {
-      log.Printf("pre-wait ProcessState=%v", cmd.ProcessState.ExitCode())
-      err = cmd.Wait()
-      log.Printf("Command finished with error: %v", err)
+	if err != nil {
+		return nil, err
+	}
 
-      // TODO mark cmd as finished in subroutine
-  }()
-  hostname, err := os.Hostname()
+	pgid, err := syscall.Getpgid(cmd.Process.Pid)
 
-  // for now make this for unique task_id
-  task_id := hostname + "-" + strconv.Itoa(cmd.Process.Pid)
-  TaskIdCmdMap[task_id] = cmd
+	if err != nil {
+		return nil, err
+	}
 
-  log.Printf("RunCommand task_id=%s, pid=%d", task_id, TaskIdCmdMap[task_id].Process.Pid)
+	taskID := hostname + "-" + strconv.Itoa(pgid) // TODO handle if Process or Pid nil
+	startResponse.TaskID = taskID
 
-  return task_id
+	taskIDCommandMap.Put(taskID, NewCommandWrapper(cmd, outBuf))
+
+	go func() {
+		// TODO add append error to CommandWrapper, impl accessors and setters
+		waitErr := cmd.Wait()
+		cw, _ := taskIDCommandMap.Get(taskID)
+		if waitErr != nil {
+			cw.SetExecError(waitErr.Error())
+		}
+		cw.SetExitCode(cmd.ProcessState.ExitCode())
+		cw.SetEndTime(time.Now())
+	}()
+
+	return &startResponse, nil
+}
+
+func intPtr(value int) *int {
+	return &value
 }
 
-// @param pid - stops process with this pid
-func StopProcess(task_id string) int {
+// StopProcess is called to end a previously started process.
+func StopProcess(taskID string) (*models.StopResponse, error) {
+
+	var stopResponse models.StopResponse
+
+	stopResponse.TaskID = taskID
+
+	var command *CommandWrapper
+	var ok bool
+
+	// if task_id invalid, return error.
+	if command, ok = taskIDCommandMap.Get(taskID); !ok {
+		return &stopResponse, errors.New("invalid task_id")
+	}
+
+	if command.GetEndTime() != nil {
+		stopResponse.ExitCode = intPtr(command.GetExitCode())
+		return &stopResponse, errors.New("process already finished")
+	}
+
+	// send stop signal first
+	// TODO verify flat processes work with gpid as well
+
+	pgid, _ := syscall.Getpgid(command.Command.Process.Pid)
+	err := syscall.Kill(-pgid, syscall.SIGQUIT)
+
+	if err != nil && err != syscall.EPERM {
```

Comment:
Can I get some clarification on this point? 

Are you talking about in the case the first SIGQUIT is sent and the error is overwritten with the subsequent SIGKILL? In that case I can add a field to CommandWrapper for quit_error. Or I can return an array of errors from the StopProcess handler.

Reply:
I think I understand what you're saying, syscall.Kill is it's own process with an output (?)

Reply:
No, what I actually meant was that the process's output seemed to not be persisted after the stop request, so inspecting the stopped process (w/o fetching status while it was running) would return empty output. I was not sure this was the intended behavior.

Reply:
`MacBook-Pro-7:grpc mymac$ go run client/*go start -command ./test_process.sh 
{"task_id":"987f769fca40-39313"}
MacBook-Pro-7:grpc mymac$ go run client/*go stop -task_id 987f769fca40-39313
{"task_id":"987f769fca40-39313","exit_code":-1}
MacBook-Pro-7:grpc mymac$ go run client/*go status -task_id 987f769fca40-39313
{"task_id":"987f769fca40-39313","start_time":"2019-09-03T23:44:22.0981362Z","end_time":"2019-09-03T23:44:30.0391807Z","exit_code":-1,"output":["I'm the script with pid 39313","Still running 39313","Still running 39313","Still running 39313","Still running 39313","Still running 39313","Still running 39313"],"execError":"signal: quit"}
`

I see what you're talking about. The status command actually queries the output directly from the CommandWrapper (Command) object, so the output does persist. Now I could change that and add a field to the CommandWrapper to set output on os.Wait completion and stop command. But at the moment, the output will still exist.

Reply:
Can't believe I've just noticed, retrieving a status consecutively eats the output the first time because it's from a buffer. I'll have to do as you suggested.